File: src/mvmctl/core/_shared/_guestfs/_kernel_detector.py
from __future__ import annotations

import logging
import re
import subprocess
from functools import lru_cache
from pathlib import Path

logger = logging.getLogger(__name__)


class KernelDetector:
    """Detects the best host kernel for libguestfs appliance builds."""

    _KERNEL_NAMES: tuple[str, ...] = (
        "vmlinuz",
        "bzImage",
        "kernel",
    )

    _DRIVER_EXTENSIONS: tuple[str, ...] = (".ko", ".ko.zst", ".ko.xz")

    _VIRTIO_NET_PATHS: tuple[Path, ...] = (Path("kernel/drivers/net"),)

    @classmethod
    @lru_cache(maxsize=1)
    def find_best_kernel(cls) -> tuple[Path, Path] | None:
        """Find the best host kernel for libguestfs.

        Scans /boot for kernel images, extracts their version strings,
        and scores candidates based on virtio module availability and
        whether the kernel looks like a custom build.

        A kernel is NOT rejected outright if virtio modules are missing
        (some distros build virtio drivers into the kernel). Instead,
        kernels with more virtio modules and without custom suffixes
        are ranked higher.

        Returns:
            Tuple of (kernel_path, modules_dir) for the best candidate,
            or None if no kernel is found.
        """
        candidates = cls._scan_boot_directory()
        if not candidates:
            return None

        scored: list[tuple[tuple[Path, Path], int]] = []
        for kernel_path, version in candidates:
            modules_dir = Path(f"/lib/modules/{version}")
            if not modules_dir.is_dir():
                logger.debug(
                    "Modules directory missing for %s: %s",
                    kernel_path,
                    modules_dir,
                )
                continue

            # Score: virtio_net bonus (critical for guestfs) + total virtio count
            virtio_net_bonus = cls._count_virtio_net(modules_dir) * 2
            virtio_count = cls._count_virtio_drivers(modules_dir)
            custom_penalty = cls._custom_suffix_penalty(version)
            score = virtio_net_bonus + virtio_count - custom_penalty
            scored.append(((kernel_path, modules_dir), score))

            logger.debug(
                "Kernel %s: virtio_net=%d total=%d penalty=%d score=%d",
                kernel_path.name,
                virtio_net_bonus // 2,
                virtio_count,
                custom_penalty,
                score,
            )

        if not scored:
            return None

        scored.sort(key=lambda item: item[1], reverse=True)
        best = scored[0][0]
        logger.debug(
            "Selected kernel %s with modules %s",
            best[0],
            best[1],
        )
        return best

    @classmethod
    def _scan_boot_directory(cls) -> list[tuple[Path, str]]:
        """Scan /boot for kernel files and extract versions."""
        boot_dir = Path("/boot")
        candidates: list[tuple[Path, str]] = []

        if not boot_dir.is_dir():
            return candidates

        paths: list[Path] = []
        for name in cls._KERNEL_NAMES:
            for path in boot_dir.glob(f"{name}*"):
                if path.is_file():
                    paths.append(path)

        for path in paths:
            version = cls._extract_version(path)
            if version:
                candidates.append((path, version))

        return candidates

    @classmethod
    def _extract_version(cls, kernel_path: Path) -> str | None:
        """Extract kernel version from a kernel image using the file command."""
        try:
            result = subprocess.run(
                ["file", str(kernel_path)],
                capture_output=True,
                text=True,
                check=False,
                timeout=5,
            )
            output = result.stdout
        except (subprocess.TimeoutExpired, FileNotFoundError, OSError) as e:
            logger.debug("Failed to run file on %s: %s", kernel_path, e)
            return None

        match = re.search(r"version\s+(\S+)", output)
        if match:
            return match.group(1)

        # Fallback: try to extract from filename for vmlinuz-X.Y.Z...
        name = kernel_path.name
        for prefix in cls._KERNEL_NAMES:
            if name.startswith(prefix):
                remainder = name[len(prefix) :]
                if remainder.startswith("-"):
                    version = remainder[1:]
                    if version and re.match(r"^\d", version):
                        return version
                break

        return None

    @classmethod
    def _count_virtio_net(cls, modules_dir: Path) -> int:
        """Count virtio_net driver files (most critical for guestfs)."""
        count = 0
        for rel_path in cls._VIRTIO_NET_PATHS:
            search_path = modules_dir / rel_path
            if not search_path.is_dir():
                continue
            for ext in cls._DRIVER_EXTENSIONS:
                count += len(list(search_path.glob(f"virtio_net{ext}")))
        return count

    @classmethod
    def _count_virtio_drivers(cls, modules_dir: Path) -> int:
        """Count all virtio drivers under the modules directory."""
        drivers_dir = modules_dir / "kernel/drivers"
        if not drivers_dir.is_dir():
            return 0

        count = 0
        for pattern in ("virtio_*.ko", "virtio_*.ko.zst", "virtio_*.ko.xz"):
            count += len(list(drivers_dir.rglob(pattern)))
        return count

    @classmethod
    def _custom_suffix_penalty(cls, version: str) -> int:
        """Return penalty for custom kernel suffixes.

        Clean versions like 6.9.3 get 0 penalty.
        Distro versions with standard suffixes like 6.9.3-1, 6.8.0-40-generic,
        7.0.2-arch1-1, 6.1.0-rc7 get 1 penalty.
        Custom versions with non-standard suffixes like -g14, -custom get 5 penalty.
        """
        import re

        # Explicit custom suffix patterns
        if re.search(r"-(g14|custom)$", version):
            return 5

        if re.match(r"^\d+\.\d+\.\d+$", version):
            return 0
        if re.match(r"^\d+\.\d+\.\d+[-.]", version):
            return 1
        return 5

File: internal/infra/guestfs/kernel_detector.go
package guestfs

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"sync"

	"mvmctl/internal/infra"
	"mvmctl/internal/infra/system"
)

// KernelDetector detects the best host kernel for libguestfs appliance builds.
// Mirrors src/mvmctl/core/_shared/_guestfs/_kernel_detector.py exactly.
type KernelDetector struct{}

var (
	kernelNames      = []string{"vmlinuz", "bzImage", "kernel"}
	driverExtensions = []string{".ko", ".ko.zst", ".ko.xz"}
	virtioNetPaths   = []string{"kernel/drivers/net"}

	// cached result for findBestKernel (like lru_cache in Python)
	cachedKernelOnce sync.Once
	cachedKernelPath string
	cachedModulesDir string
	cachedKernelErr  error
)

// FindBestKernel finds the best host kernel for libguestfs.
// Returns (kernelPath, modulesDir, error).
// Scans /boot for kernel images, extracts version strings, and scores
// candidates based on virtio module availability and custom build penalty.
func (kd *KernelDetector) FindBestKernel(ctx context.Context) (string, string, error) {
	cachedKernelOnce.Do(func() {
		cachedKernelPath, cachedModulesDir, cachedKernelErr = kd.findBestKernelUncached(ctx)
	})
	return cachedKernelPath, cachedModulesDir, cachedKernelErr
}

func (kd *KernelDetector) findBestKernelUncached(ctx context.Context) (string, string, error) {
	candidates, err := kd.scanBootDirectory(ctx)
	if err != nil {
		return "", "", err
	}
	if len(candidates) == 0 {
		return "", "", nil
	}

	type scored struct {
		kernelPath string
		modulesDir string
		score      int
	}

	var scoredCandidates []scored

	for _, c := range candidates {
		kernelPath, version := c.kernelPath, c.version
		modulesDir := filepath.Join("/lib/modules", version)
		info, err := os.Stat(modulesDir)
		if err != nil || !info.IsDir() {
			slog.Debug("Modules directory missing for kernel",
				"kernel", kernelPath,
				"modules", modulesDir,
			)
			continue
		}

		virtioNetBonus := kd.countVirtioNet(modulesDir) * 2
		virtioCount := kd.countVirtioDrivers(modulesDir)
		customPenalty := kd.customSuffixPenalty(version)
		score := virtioNetBonus + virtioCount - customPenalty

		scoredCandidates = append(scoredCandidates, scored{
			kernelPath: kernelPath,
			modulesDir: modulesDir,
			score:      score,
		})

		slog.Debug("Kernel scored",
			"name", filepath.Base(kernelPath),
			"virtio_net", virtioNetBonus/2,
			"total", virtioCount,
			"penalty", customPenalty,
			"score", score,
		)
	}

	if len(scoredCandidates) == 0 {
		return "", "", nil
	}

	sort.Slice(scoredCandidates, func(i, j int) bool {
		return scoredCandidates[i].score > scoredCandidates[j].score
	})

	best := scoredCandidates[0]
	slog.Debug("Selected kernel",
		"kernel", best.kernelPath,
		"modules", best.modulesDir,
	)
	return best.kernelPath, best.modulesDir, nil
}

type kernelCandidate struct {
	kernelPath string
	version    string
}

func (kd *KernelDetector) scanBootDirectory(ctx context.Context) ([]kernelCandidate, error) {
	bootDir := "/boot"
	info, err := os.Stat(bootDir)
	if err != nil || !info.IsDir() {
		return nil, nil
	}

	var paths []string
	for _, name := range kernelNames {
		matches, err := filepath.Glob(filepath.Join(bootDir, name+"*"))
		if err != nil {
			continue
		}
		for _, m := range matches {
			fi, err := os.Stat(m)
			if err == nil && !fi.IsDir() {
				paths = append(paths, m)
			}
		}
	}

	// Deduplicate paths while preserving order
	unique := infra.Dedup(paths)

	var candidates []kernelCandidate
	for _, p := range unique {
		version, verErr := kd.extractVersion(ctx, p)
		if verErr != nil {
			return nil, verErr
		}
		if version != "" {
			candidates = append(candidates, kernelCandidate{
				kernelPath: p,
				version:    version,
			})
		}
	}
	return candidates, nil
}

func (kd *KernelDetector) extractVersion(ctx context.Context, kernelPath string) (string, error) {
	// Try 'file' command first
	// Python's _extract_version: raises ProcessError on timeout,
	// logs and returns None on other errors.
	result := system.RunCmdCompat(ctx, []string{"file", kernelPath}, system.RunCmdOpts{
		Capture: true,
		Check:   false,
		Timeout: 5000,
	})
	if result.Err != nil {
		// Check for timeout — re-raise with specific message like Python
		if strings.Contains(result.Err.Error(), "timed out") {
			return "", fmt.Errorf("'file' command timed out for %s", kernelPath)
		}
		// Other errors: silently fall through to filename fallback
	} else if result.Stdout != "" {
		re := regexp.MustCompile(`version\s+(\S+)`)
		matches := re.FindStringSubmatch(result.Stdout)
		if len(matches) >= 2 {
			return matches[1], nil
		}
	}

	// Fallback: extract from filename for vmlinuz-X.Y.Z...
	name := filepath.Base(kernelPath)
	for _, prefix := range kernelNames {
		if strings.HasPrefix(name, prefix) {
			remainder := name[len(prefix):]
			if strings.HasPrefix(remainder, "-") {
				version := remainder[1:]
				if version != "" {
					matched, _ := regexp.MatchString(`^\d`, version)
					if matched {
						return version, nil
					}
				}
			}
			break
		}
	}

	return "", nil
}

func (kd *KernelDetector) countVirtioNet(modulesDir string) int {
	count := 0
	for _, relPath := range virtioNetPaths {
		searchPath := filepath.Join(modulesDir, relPath)
		info, err := os.Stat(searchPath)
		if err != nil || !info.IsDir() {
			continue
		}
		for _, ext := range driverExtensions {
			pattern := filepath.Join(searchPath, "virtio_net"+ext)
			matches, _ := filepath.Glob(pattern)
			count += len(matches)
		}
	}
	return count
}

func (kd *KernelDetector) countVirtioDrivers(modulesDir string) int {
	driversDir := filepath.Join(modulesDir, "kernel/drivers")
	info, err := os.Stat(driversDir)
	if err != nil || !info.IsDir() {
		return 0
	}

	// Build set of matching extensions for fast lookup
	extSet := make(map[string]bool)
	for _, ext := range driverExtensions {
		extSet[ext] = true
	}

	count := 0
	filepath.WalkDir(driversDir, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return nil // skip errors
		}
		if d.IsDir() {
			return nil
		}
		name := d.Name()
		ext := filepath.Ext(name)
		if !extSet[ext] {
			return nil
		}
		// Check if name starts with "virtio_"
		baseName := name[:len(name)-len(ext)]
		if strings.HasPrefix(baseName, "virtio_") {
			count++
		}
		return nil
	})
	return count
}

func (kd *KernelDetector) customSuffixPenalty(version string) int {
	// Explicit custom suffix patterns: -g14, -custom at end
	matchedCustom, _ := regexp.MatchString(`-(g14|custom)$`, version)
	if matchedCustom {
		return 5
	}

	// Clean version like 6.9.3
	matchedClean, _ := regexp.MatchString(`^\d+\.\d+\.\d+$`, version)
	if matchedClean {
		return 0
	}

	// Distro version like 6.9.3-1, 6.8.0-40-generic
	matchedDistro, _ := regexp.MatchString(`^\d+\.\d+\.\d+[-.]`, version)
	if matchedDistro {
		return 1
	}

	return 5
}

// ResetCache clears the cached kernel find result. Used for testing.
func (kd *KernelDetector) ResetCache() {
	cachedKernelOnce = sync.Once{}
	cachedKernelPath = ""
	cachedModulesDir = ""
	cachedKernelErr = nil
}

// Ensure kernelDetector satisfies the interface.
var _ = (*KernelDetector)(nil)

// For backward compatibility and to match Python's classmethod pattern.
func FindBestKernel(ctx context.Context) (string, string, error) {
	kd := &KernelDetector{}
	return kd.FindBestKernel(ctx)
}

func ScanBootDirectory(ctx context.Context) ([]string, error) {
	kd := &KernelDetector{}
	candidates, err := kd.scanBootDirectory(ctx)
	if err != nil {
		return nil, err
	}
	result := make([]string, len(candidates))
	for i, c := range candidates {
		result[i] = c.kernelPath
	}
	return result, nil
}
